File: src/windows/AnyTalk/Services/WhisperService.cs
using System.Net.Http.Headers;
using System.Text.Json;

namespace AnyTalk.Services;

public enum WhisperError
{
    NoAPIKey,
    InvalidAudioFile,
    NetworkError,
    APIError
}

public class WhisperService
{
    private static WhisperService? instance;
    private readonly HttpClient httpClient;

    public static WhisperService Instance
    {
        get
        {
            instance ??= new WhisperService();
            return instance;
        }
    }

    private WhisperService()
    {
        httpClient = new HttpClient();
    }

    public async Task<Result<string>> TranscribeAudio(string audioFilePath)
    {
        var settings = SettingsManager.Instance.LoadSettings();
        var apiKey = settings.ApiKey;

        if (string.IsNullOrEmpty(apiKey))
        {
            return Result<string>.Failure(WhisperError.NoAPIKey, "API key is required");
        }

        // Check if the audio file exists
        if (!File.Exists(audioFilePath))
        {
            return Result<string>.Failure(WhisperError.InvalidAudioFile, "Audio file not found");
        }

        try
        {
            var request = new HttpRequestMessage(HttpMethod.Post, "https://api.openai.com/v1/audio/transcriptions");
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);

            var boundary = Guid.NewGuid().ToString();
            var content = new MultipartFormDataContent(boundary);

            // Add the model parameter
            content.Add(new StringContent("whisper-1"), "model");

            // Add language parameter if user has set a preferred language
            if (!string.IsNullOrEmpty(settings.Language) && settings.Language != "auto")
            {
                content.Add(new StringContent(settings.Language), "language");
            }

            // Set a lower temperature for more deterministic results
            content.Add(new StringContent("0.3"), "temperature");

            // Add the audio file
            var fileContent = new ByteArrayContent(File.ReadAllBytes(audioFilePath));
            fileContent.Headers.ContentType = MediaTypeHeaderValue.Parse("audio/m4a");
            content.Add(fileContent, "file", Path.GetFileName(audioFilePath));

            request.Content = content;

            var response = await httpClient.SendAsync(request);
            var responseBody = await response.Content.ReadAsStringAsync();

            if (!response.IsSuccessStatusCode)
            {
                var errorResponse = JsonSerializer.Deserialize<ErrorResponse>(responseBody);
                var errorMessage = errorResponse?.Error?.Message ?? $"API error with status code: {response.StatusCode}";
                return Result<string>.Failure(WhisperError.APIError, errorMessage);
            }

            var result = JsonSerializer.Deserialize<WhisperResponse>(responseBody);
            if (string.IsNullOrEmpty(result?.Text))
            {
                return Result<string>.Failure(WhisperError.APIError, "Could not parse response");
            }

            return Result<string>.Success(result.Text);
        }
        catch (Exception ex)
        {
            return Result<string>.Failure(WhisperError.NetworkError, ex.Message);
        }
    }

    private class WhisperResponse
    {
        public string Text { get; set; } = string.Empty;
    }

    private class ErrorResponse
    {
        public ErrorDetail? Error { get; set; }
    }

    private class ErrorDetail
    {
        public string Message { get; set; } = string.Empty;
    }
}

public class Result<T>
{
    public bool IsSuccess { get; private set; }
    public T? Value { get; private set; }
    public WhisperError Error { get; private set; }
    public string ErrorMessage { get; private set; }

    private Result(bool isSuccess, T? value, WhisperError error, string errorMessage)
    {
        IsSuccess = isSuccess;
        Value = value;
        Error = error;
        ErrorMessage = errorMessage;
    }

    public static Result<T> Success(T value)
    {
        return new Result<T>(true, value, default, string.Empty);
    }

    public static Result<T> Failure(WhisperError error, string message)
    {
        return new Result<T>(false, default, error, message);
    }
}

File: AnyTalk/Services/WhisperService.swift
import Foundation
import AVFoundation

enum WhisperError: Error {
    case noAPIKey
    case invalidAudioFile
    case networkError(String)
    case apiError(String)
}

class WhisperService {
    static let shared = WhisperService()
    
    private init() {}
    
    func transcribe(audioURL: URL, completion: @escaping (Result<String, Error>) -> Void) {
        let apiKey = SettingsManager.shared.apiKey
        guard !apiKey.isEmpty else {
            completion(.failure(WhisperError.noAPIKey))
            return
        }
        
        guard FileManager.default.fileExists(atPath: audioURL.path) else {
            completion(.failure(WhisperError.invalidAudioFile))
            return
        }
        
        let url = URL(string: "https://api.openai.com/v1/audio/transcriptions")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        
        let boundary = UUID().uuidString
        request.addValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        
        let httpBody = NSMutableData()
        
        httpBody.append("--\(boundary)\r\n".data(using: .utf8)!)
        httpBody.append("Content-Disposition: form-data; name=\"model\"\r\n\r\n".data(using: .utf8)!)
        httpBody.append("whisper-1\r\n".data(using: .utf8)!)
        
        let language = SettingsManager.shared.preferredLanguage
        if !language.isEmpty && language != "auto" {
            httpBody.append("--\(boundary)\r\n".data(using: .utf8)!)
            httpBody.append("Content-Disposition: form-data; name=\"language\"\r\n\r\n".data(using: .utf8)!)
            httpBody.append("\(language)\r\n".data(using: .utf8)!)
        }
        
        httpBody.append("--\(boundary)\r\n".data(using: .utf8)!)
        httpBody.append("Content-Disposition: form-data; name=\"temperature\"\r\n\r\n".data(using: .utf8)!)
        httpBody.append("0.3\r\n".data(using: .utf8)!)
        
        let promptText = "Return text with proper grammar, punctuation, and formatting. Direct speech should be in quotation marks. Format numbers and dates properly."
        httpBody.append("--\(boundary)\r\n".data(using: .utf8)!)
        httpBody.append("Content-Disposition: form-data; name=\"prompt\"\r\n\r\n".data(using: .utf8)!)
        httpBody.append("\(promptText)\r\n".data(using: .utf8)!)

        httpBody.append("--\(boundary)\r\n".data(using: .utf8)!)
        httpBody.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(audioURL.lastPathComponent)\"\r\n".data(using: .utf8)!)
        httpBody.append("Content-Type: audio/m4a\r\n\r\n".data(using: .utf8)!)
        
        if let audioData = try? Data(contentsOf: audioURL) {
            httpBody.append(audioData)
            httpBody.append("\r\n".data(using: .utf8)!)
        } else {
            completion(.failure(WhisperError.invalidAudioFile))
            return
        }
        
        httpBody.append("--\(boundary)--\r\n".data(using: .utf8)!)
        
        request.httpBody = httpBody as Data
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                DispatchQueue.main.async {
                    completion(.failure(WhisperError.networkError(error.localizedDescription)))
                }
                return
            }
            
            guard let data = data else {
                DispatchQueue.main.async {
                    completion(.failure(WhisperError.networkError("No data received")))
                }
                return
            }
            
            do {
                if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode != 200 {
                    if let errorResponse = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any],
                       let errorMessage = errorResponse["error"] as? [String: Any],
                       let message = errorMessage["message"] as? String {
                        DispatchQueue.main.async {
                            completion(.failure(WhisperError.apiError(message)))
                        }
                    } else {
                        DispatchQueue.main.async {
                            completion(.failure(WhisperError.apiError("API error with status code: \(httpResponse.statusCode)")))
                        }
                    }
                    return
                }
                
                let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
                
                if let text = json?["text"] as? String {
                    DispatchQueue.main.async {
                        completion(.success(text))
                    }
                } else {
                    DispatchQueue.main.async {
                        completion(.failure(WhisperError.apiError("Could not parse response")))
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    completion(.failure(error))
                }
            }
        }
        
        task.resume()
    }
} 
